Fakebook.Profile/Fakebook.Profile.DataAccess/ProfileRepository.cs

```diff
@@ -69,75 +69,77 @@ private static DomainProfile ToDomainProfile(EntityProfile profile)
         /// Default model mapping from domain to entity.
         /// </summary>
         /// <param name="profile">domain model object used</param>
-        /// <exception type="ArgumentNullException">If the profile, the profile's email, or the profile's names are null,
+        /// <exception type="ArgumentException">If the profile, the profile's email, or the profile's names are null,
         /// this will be thrown </exception>
         /// <returns>A representation of the profile as a DB Entity.</returns>
-        private Task<EntityProfile> ToEntityProfile(DomainProfile profile)
+        private async Task<EntityProfile> ToEntityProfileAsync(DomainProfile profile)
         {
             if (profile == null || profile.Email == null || profile.FirstName == null || profile.LastName == null)
             {
-                throw new ArgumentNullException("Must have a domain profile, with an email and first and last name.");
+                throw new ArgumentException("Profile is missing required fields.", nameof(profile));
             }
 
-            return ToEntityProfileAsync(profile);
-        }
+            var entityProfile = await ConvertAsync(profile);
+            return entityProfile;
 
-        private async Task<EntityProfile> ToEntityProfileAsync(DomainProfile profile) {
+            async Task<EntityProfile> ConvertAsync(DomainProfile profile)
+            {
 
-            List<Follow> followingEmails = new List<Follow>();
-            List<Follow> followerEmails = new List<Follow>();
+                List<Follow> followingEmails = new List<Follow>();
+                List<Follow> followerEmails = new List<Follow>();
 
-            if(profile.FollowingEmails.Count != 0)
-            {
-                int userId = await GetProfileIdAsync(profile.Email);
-                foreach (var following in profile.FollowingEmails)
+                if (profile.FollowingEmails.Count != 0)
                 {
-                    int followingId = await GetProfileIdAsync(following);
-                    Follow newFollow = new Follow
+                    int userId = await GetProfileIdAsync(profile.Email);
+                    foreach (var following in profile.FollowingEmails)
                     {
-                        UserId = userId,
-                        FollowingId = followingId
-                    };
-                    followingEmails.Add(newFollow);
+                        int followingId = await GetProfileIdAsync(following);
```

Comment:
`await AccessDatabaseAsync();` inside a loop is pretty bad, since we have to wait the whole network roundtrip for each query. instead, construct a query that gets all the info you want in one go. likewise for below.

Reply:
Made a new method to get many Ids from a list of emails to avoid this issue.